File: 백준/Silver/24392. 영재의 징검다리/영재의 징검다리.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int MOD = (int)1e9 + 7;

        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        int[][] board = new int[n][m];
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < m; j++)
                board[i][j] = Integer.parseInt(st.nextToken());
        }

        int[] now = board[0].clone();
        for (int i = 1; i < n; i++) {
            int[] nxt = new int[m];
            for (int j = 0; j < m; j++) {
                if (board[i][j] == 0)
                    continue;

                nxt[j] = ((now[j] + (j > 0 ? now[j-1] : 0)) % MOD + (j < m-1 ? now[j+1] : 0)) % MOD;

            }
            now = nxt;
        }

        int total = 0;
        for (int j = 0; j < m; j++) {
            total = (total + now[j]) % MOD;
        }
        System.out.println(total);
    }
}

File: 백준/Silver/24392. 영재의 징검다리/영재의 징검다리.py
import sys
input = sys.stdin.readline
MOD = int(1e9) + 7

n,  m = map(int, input().split())
board = [list(map(int, input().split())) for _ in range(n)]

now = board[-1].copy() + [0]
for i in range(n-2, -1, -1):
    nxt = [0] * (m+1)
    for j in range(m):
        nxt[j] = (now[j-1] + now[j] + now[j+1]) % MOD if board[i][j] else 0

    now = nxt

print(sum(now) % MOD)